storage/store.go

```diff
@@ -175,42 +193,27 @@ func (e *NotBootstrappedError) Error() string {
 	return "store has not been bootstrapped"
 }
 
-// storeRangeIterator is an implementation of rangeIterator which
-// cycles through a store's rangesByKey slice.
-type storeRangeIterator struct {
-	store     *Store
-	remaining int
-	index     int
-}
-
-func newStoreRangeIterator(store *Store) *storeRangeIterator {
-	r := &storeRangeIterator{
-		store: store,
-	}
-	r.Reset()
-	return r
+// storeRangeSet is an implementation of rangeSet which
+// cycles through a store's rangesByKey btree with mutex locking.
+type storeRangeSet struct {
+	store *Store
 }
 
-func (si *storeRangeIterator) Next() *Range {
-	si.store.mu.Lock()
-	defer si.store.mu.Unlock()
-	if index, remaining := si.index, len(si.store.rangesByKey)-si.index; remaining > 0 {
-		si.index++
-		si.remaining = remaining - 1
-		return si.store.rangesByKey[index]
-	}
-	return nil
-}
-
-func (si *storeRangeIterator) EstimatedCount() int {
-	return si.remaining
+func (rs *storeRangeSet) Visit(iterator func(*Range) bool) {
+	// Hold the lock except when the iterator is called.
```

Comment:
Sure, pushed a commit (59d400869b3d2e55008b80d960b9ead3d4fe2ca8) that changes the `Visit` method to make a copy.

The commit tweaks `TestStoreRangeIterator ` for testing. I've also fixed a bug of `EstimatedCount()` in my previous commit. (The method needs to return the remaining number of ranges.)

@bdarnell , @tamird , could you take another look?